File: javascript/packages/linter/src/rules/html-aria-attribute-must-be-valid.ts
import { ParserRule } from "../types.js";
import { ARIA_ATTRIBUTES, AttributeVisitorMixin } from "./rule-utils.js";

import type { LintOffense, LintContext } from "../types.js";
import type {HTMLAttributeNode, HTMLOpenTagNode, HTMLSelfCloseTagNode, Node, ParseResult } from "@herb-tools/core";

class AriaAttributeMustBeValid extends AttributeVisitorMixin {
  protected checkStaticAttributeStaticValue(
    attributeName: string, _attributeValue: string, attributeNode: HTMLAttributeNode, _parentNode: HTMLOpenTagNode | HTMLSelfCloseTagNode): void {
    if (!attributeName.startsWith("aria-")) return;

    if (!ARIA_ATTRIBUTES.has(attributeName)){
      this.addOffense(
        `The attribute \`${attributeName}\` is not a valid ARIA attribute. ARIA attributes must match the WAI-ARIA specification.`,
        attributeNode.location,
        "error"
      );
    }
  }

  protected checkStaticAttributeDynamicValue(attributeName: string, _valueNodes: Node[], attributeNode: HTMLAttributeNode, _parentNode: HTMLOpenTagNode | HTMLSelfCloseTagNode, _combinedValue?: string): void {
    if (!attributeName.startsWith("aria-")) return;

    if (!ARIA_ATTRIBUTES.has(attributeName)){
      this.addOffense(
        `The attribute \`${attributeName}\` is not a valid ARIA attribute. ARIA attributes must match the WAI-ARIA specification.`,
        attributeNode.location,
        "error"
      );
    }
  }
}

export class HTMLAriaAttributeMustBeValid extends ParserRule {
  name = "html-aria-attribute-must-be-valid";

  check(result: ParseResult, context?: Partial<LintContext>): LintOffense[] {
    const visitor = new AriaAttributeMustBeValid(this.name, context);

    visitor.visit(result.value);

    return visitor.offenses;
  }
}

File: rust/herb-linter/src/rules/html_aria_attribute_must_be_valid.rs
use crate::utils::aria_data::ARIA_ATTRIBUTES;
use crate::utils::tag_utils::get_attribute_name;

use herb::nodes::{AnyNode, ERBOpenTagNode, HTMLOpenTagNode};
use herb::Visitor;

rule_visitor!(AriaAttributeMustBeValidVisitor);
define_parser_rule!(
  HTMLAriaAttributeMustBeValidRule,
  "html-aria-attribute-must-be-valid",
  Warning,
  AriaAttributeMustBeValidVisitor,
  parser_options: { action_view_helpers: true },
  introduced_in: "0.4.1"
);

impl AriaAttributeMustBeValidVisitor {
  fn check_children(&mut self, children: &[AnyNode]) {
    for attribute in children.iter().filter_map(|child| match child {
      AnyNode::HTMLAttributeNode(attribute) => Some(attribute.as_ref()),
      _ => None,
    }) {
      if let Some(name) = get_attribute_name(attribute) {
        let lowercase_name = name.to_lowercase();

        if !lowercase_name.starts_with("aria-") {
          continue;
        }

        if ARIA_ATTRIBUTES.contains(lowercase_name.as_str()) {
          continue;
        }

        self.add_offense(
          format!(
            "The attribute `{}` is not a valid ARIA attribute. ARIA attributes must match the WAI-ARIA specification.",
            lowercase_name
          ),
          attribute.location.clone(),
        );
      }
    }
  }
}

impl Visitor for AriaAttributeMustBeValidVisitor {
  fn visit_html_open_tag_node(&mut self, node: &HTMLOpenTagNode) {
    self.check_children(&node.children);
    self.walk_html_open_tag_node(node);
  }

  fn visit_erb_open_tag_node(&mut self, node: &ERBOpenTagNode) {
    self.check_children(&node.children);
    self.walk_erb_open_tag_node(node);
  }
}
